Add api command that pauses alerts belonging to a single dashboard

Jonnymcc opened the pull request:
Related to #6500.

Commit:
Add api command that pauses alerts belonging to a single dashboard

bergquist commented on pkg/api/api.go:
```diff
@@ -335,6 +335,7 @@ func (hs *HttpServer) registerRoutes() {
 		adminRoute.Get("/users/:id/quotas", wrap(GetUserQuotas))
 		adminRoute.Put("/users/:id/quotas/:target", bind(m.UpdateUserQuotaCmd{}), wrap(UpdateUserQuota))
 		adminRoute.Get("/stats", AdminGetStats)
+		adminRoute.Post("/:dashboardId/pause-dashboard-alerts", bind(dtos.PauseDashboardAlertsCommand{}), wrap(PauseDashboardAlerts))
```
I don't think this should be under `/admin`. We already have a route there but I don't it should be there. 
I think it's better to have it under `/api/alerts` https://github.com/grafana/grafana/pull/10622/files#diff-13d216b12482e706601d65bcc87deb38R293

maybe
POST `/api/alerts/pause/by-dashboard/:dashboardId`

not :100: sure yet...

Jonnymcc replied:
And then change the others to
`/api/alerts/pause/by-id/:alertId`
`/api/alerts/pause/all`
?
I found it confusing that `pause-all-alerts` was under the admin route and not the alerts route.

bergquist replied:
I don't want to change the `pause-all-alerts` path since that would be a breaking change for everyone who wrote scripts to do that. But I think we could have two paths do the same thing.

bergquist commented on pkg/services/sqlstore/alert.go:
```diff
@@ -295,6 +296,34 @@ func PauseAlert(cmd *m.PauseAlertCommand) error {
+		}
+
+		var buffer bytes.Buffer
+		params := make([]interface{}, 0)
```
Avoid using make for creating empty slices. Instead, create an empty slice

```go
var params []interface{}
``` 

https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

bergquist commented on pkg/services/sqlstore/alert.go:
```diff
@@ -295,6 +296,34 @@ func PauseAlert(cmd *m.PauseAlertCommand) error {
+			return fmt.Errorf("command contains no dashboard id")
+		}
+
+		var buffer bytes.Buffer
```
I dont think we need to create a Buffer for this. A more clear way would be to just have one sql statement. 

`sql := "UPDATE alert SET state = ? WHERE dashboard_id = ?`

bergquist reviewed (changes requested):
Great! Looks good! But, I would like to see some minor improvements before we merge this.

Commit:
Move dashboardId check to api func instead of sqlstore func

Commit:
Simplify var declaration/initialization

Commit:
Replace usage of buffer with simple string declaration

Commit:
Create empty slice without using make

Commit:
Move pause alert by dashboard to /alerts api route group

Commit:
Add reqEditorRole, to pause alert by dashboard api route

Commit:
Add /pause/by-id/:alertId route for consistency

Jonnymcc commented:
@bergquist I think I addressed all of your change requests. I also added the `reqEditorRole` to the route and added another route `/pause/by-id/:alertId` for consistency.

bergquist commented:
Great! I'll review it as soon as I have time :)

bergquist commented on pkg/api/api.go:
```diff
@@ -293,6 +293,8 @@ func (hs *HttpServer) registerRoutes() {
 			alertsRoute.Post("/test", bind(dtos.AlertTestCommand{}), wrap(AlertTest))
 			alertsRoute.Post("/:alertId/pause", reqEditorRole, bind(dtos.PauseAlertCommand{}), wrap(PauseAlert))
+			alertsRoute.Post("/pause/by-id/:alertId", reqEditorRole, bind(dtos.PauseAlertCommand{}), wrap(PauseAlert))
+			alertsRoute.Post("/pause/by-dashboard/:dashboardId", reqEditorRole, bind(dtos.PauseDashboardAlertsCommand{}), wrap(PauseDashboardAlerts))
```
I'm still not sure about this path. (I know it was my suggestion.. )

We want to support pausing by dashboardid, datasource and dashboard folder (unreleased feature).
Having one endpoint for each seems messy. So I think we can combine them all into one endpoint and get the dashboardid/datasourceid/folderid from the json body instead.

`POST /alerts/pause` should require at least one of those fields to be set and pause alert rules according to the provided fields. 

@torkelo any thought on this? So far this is the only time we update entities based on foreign Id. I'm not 100% sure how we want to do that. 

All other changes look good. Thank you for being so patient with my feedback :)

Jonnymcc replied:
I suppose specifying the id in the json body also makes it possible to send a list of ids and do a bulk update. This isn't possible by extracting from a URL param right?

bergquist replied:
@Jonnymcc yea that would also be a good addition.

Jonnymcc commented:
/me waiting for input form @torkelo

bergquist commented:
@Jonnymcc If you want to, go ahead and implement one POST route for handling both pausing alerts by alertId and dashboardId. Making it an array of alertIds and dashboardsIds would be even better. But you can keep the current `/api/:alertId/pause` route to avoid a breaking change.

bergquist commented:
Closing this due to inactivity. Please open a new PR if you want to take a swing at this again!
